In [1]:
import numpy as np

In [2]:
def combinations(n, k) -> int:
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

In [3]:
def bernulli(n, k, p):
    return combinations(n, k) * (p**k) * (1 - p)**(n-k)

#### Задача 1

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

Посчитаем закон распределения по формуле Бернулли:
$$P(X = k) = C_n^k p^k (1 - p)^{n - k}$$

Возможные значения исходов 0 - что ни разу не угадан, 5 - угаданы все ответы: $x_1 = 0$, $x_2 = 1$, $x_3 = 2$, $x_4 = 3$, $x_5 = 4$, $x_6 = 5$.

In [4]:
n = 5 #число вопросов
p = 1/4 #вероятность угадать правильный ответ в каждом из вопросов
i = n + 1 #число возможных исходов

for k in range(0, i):
    print(k, bernulli(n, k, p))

0 0.2373046875
1 0.3955078125
2 0.263671875
3 0.087890625
4 0.0146484375
5 0.0009765625


Итак, закон распределения такой случайной величины:

<table border="3">
    <tr>
        <th>$x$</th>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
        <td>5</td>
    </tr>
    <tr>
        <th>$P(X = x)$</th>
        <td>0.237</td>
        <td>0.396</td>
        <td>0.264</td>
        <td>0.088</td>
        <td>0.015</td>
        <td>0.001</td>
    </tr>
</table>

Вычислим математическое ожидание и дисперсию по формулам:
$$M(X) = np, \: D(X) = np(1 - p).$$

In [5]:
m_x = n * p
m_x

1.25

In [6]:
d_x = n * p * (1 - p)
d_x

0.9375

#### Задача 2

Пользователь получает в среднем 10 писем со спамом на свой почтовый ящик за сутки. Найти число N, такое, что с вероятностью 0.95 пользователь получит не более N писем со спамом за текущий день.

Применим формулу Пуассона:
$$P(X = k) = \dfrac{\lambda^k e^{-\lambda}}{k!}$$

In [7]:
def poisson_proba(k, lambda_) -> float:
    """Формула Пуассона.
    """
    
    return (lambda_ ** k) * (np.exp(-lambda_)) / np.math.factorial(k)

In [8]:
lambda_ = 10 #среднее кол-во писем со спамом

Просуммируем вероятности получения от 0 до N писем. Сложим суммы вероятностей и узнаем N, при котором вероятность = 95% (до второго знака), что пользователь получить от 0 до N писем.

In [9]:
p = 0
i = 0
while p < 0.95:
    poisson = poisson_proba(i, lambda_)
    p += poisson
#     print(i, round(poisson, 2), round(p, 2))
    i += 1

print("Ответ N =", i-1)

Ответ N = 15


#### Задача 3

Производятся выстрелы по мишени. Вероятность попадания в мишень при одном выстреле равна 0.01. Сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 6 раз будет совершено попадание?

Подсказка. Здесь предстоит немножко покодить.

In [10]:
p = 0.01 #вероятность попадания в мишень при одном выстреле
p6 = 0.9 #вероятность попадания в мишень 6 и более раз
x = 6 #минимальное кол-во раз, сколько нужно попасть в мишень

Получается, что нам дано все необходимое для формулы Бернулли.
Нам надо узнать n, при которых мы попадем в цель более 6 раз. Или по-другому, надо узнать n, при которых мы попадем в цель от 0 до 5 раз.

$$P(k > 5) = 1 - P(k < 6)$$

In [11]:
p_k_5 = 0
n = x
while p_k_5 < (1 - p6):
    be_sum = 0
    for k in range(x):
        be = bernulli(n, k, p)
        be_sum += be
#         print(n, k, round(be, 5), round(be_sum, 3))
    p_k_5 = 1 - be_sum
    n += 1

print('Ответ N =', n-1)

Ответ N = 317
